compiler/optimizations/forallOptimizations.cpp

```diff
@@ -931,7 +931,18 @@ static void gatherForallInfo(ForallStmt *forall) {
 static bool loopHasValidInductionVariables(ForallStmt *forall) {
   // if we understand the induction variables, then we can still take a look at
   // the loop body for some calls that we can optimize dynamically
-  return forall->optInfo.multiDIndices.size() > 0;
+  // if multiDIndicesSize is a different length than the other sym lists,
+  // its likely we didn't understand one of the calls (possible use-before-def)
+
+  auto multiDIndicesSize = forall->optInfo.multiDIndices.size();
+  auto iterSymSize = forall->optInfo.iterSym.size();
+  auto dotDomIterSymSize = forall->optInfo.dotDomIterSym.size();
+  auto iterCallTmpSize = forall->optInfo.iterCallTmp.size();
+
+  return multiDIndicesSize > 0 &&
+         multiDIndicesSize == iterSymSize &&
+         multiDIndicesSize == dotDomIterSymSize &&
+         iterCallTmpSize == multiDIndicesSize;
```

Comment:
Nit: swap order.

Non-nit: I am not really understanding how this is helping. Do. you remember which of these comparison help in this particular case?

Reply:
(By looking at the code, I expect all of these to have the same size always)

Reply:
This helps because of the loop in `gatherForallInfo`, which goes through all iterated expressions. It always appends to `iterSym`, `dotDomIterSym`, and `iterCallTmp`, but it only appends to `multiDIndicesSize` when it finds a non-null iter expr. In this case of a use-before-def, the mismatch is because it cannot find a symbol to put in `iterSym`.

So the check could be simplified to just `multiDIndicesSize > 0 && multiDIndicesSize == iterSymSize`. Although I think I will factor it out to a separate check as you suggested

Reply:
Hmm, that's strange (an issue with the iterand throwing the induction variable analysis off), but no need to investigate further. Thanks!

Yeah, I agree, any check you deem appropriate can go in a separate function whose failure results in a better log message.